// @whop-apps/api/dist/notifications.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "notifications", {
    enumerable: true,
    get: function() {
        return notifications;
    }
});
const _api = require("./api");
function notifications(...args) {
    const api = (0, _api.app)(...args);
    return {
        async send (notificationRequest) {
            const result = await api.POST("/app/notifications/requests", {
                body: notificationRequest
            });
            const data = result.unwrap();
            return data;
        },
        async createRequest (notificationRequest) {
            // This is done so that we automatically trim text content and don't send huge payloads to the API
            const truncatedNotificationRequest = {
                ...notificationRequest,
                body: {
                    title: notificationRequest.body.title.slice(0, 200),
                    subtitle: notificationRequest.body.subtitle?.slice(0, 200),
                    content: notificationRequest.body.content.slice(0, 200)
                }
            };
            const result = await api.POST("/app/notification_requests", {
                body: truncatedNotificationRequest
            });
            const data = result.unwrap();
            return data;
        }
    };
}
