// @whop-apps/api/dist/search.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "search", {
    enumerable: true,
    get: function() {
        return search;
    }
});
const _api = require("./api");
function search(...args) {
    const api = (0, _api.app)(...args);
    return {
        async createDocument (document) {
            const result = await api.POST("/app/search/documents", {
                body: document
            });
            const data = result.unwrap();
            return data;
        },
        async updateDocument (documentId, document) {
            const result = await api.PATCH("/app/documents/{id}", {
                params: {
                    path: {
                        id: documentId
                    }
                },
                body: document
            });
            const data = result.unwrap();
            return data;
        },
        async deleteDocument (documentId) {
            const result = await api.DELETE("/app/documents/{id}", {
                params: {
                    path: {
                        id: documentId
                    }
                }
            });
            const data = result.unwrap();
            return data;
        }
    };
}
