// @whop-apps/iframe/dist/apps-server.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "appsServerSchema", {
    enumerable: true,
    get: function() {
        return appsServerSchema;
    }
});
const _zod = require("zod");
const _utils = require("./utils");
const appsServerSchema = _zod.z.discriminatedUnion("event", [
    _zod.z.object({
        event: _zod.z.literal("appPing"),
        request: _zod.z.literal("app_ping"),
        response: _zod.z.literal("app_pong")
    }),
    _zod.z.object({
        event: _zod.z.literal("onCheckoutRequirementComplete"),
        request: _zod.z.object({
            checkoutSessionId: _zod.z.string(),
            planId: _zod.z.string(),
            productId: _zod.z.string(),
            companyId: _zod.z.string()
        }),
        response: _zod.z.object({
            shouldContinue: _zod.z.boolean(),
            error: _zod.z.string().optional()
        })
    }),
    _zod.z.object({
        event: _zod.z.literal("onColorThemeChange"),
        request: _utils.frostedV2Theme,
        response: _zod.z.void()
    })
]);
