// @whop-apps/iframe/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    withError: function() {
        return withError;
    },
    frostedV2Theme: function() {
        return frostedV2Theme;
    }
});
const _zod = require("zod");
const withError = (schema, error)=>{
    return _zod.z.discriminatedUnion("status", [
        _zod.z.object({
            status: _zod.z.literal("ok"),
            data: schema
        }),
        _zod.z.object({
            status: _zod.z.literal("error"),
            error: error
        })
    ]);
};
const frostedV2Theme = _zod.z.object({
    appearance: _zod.z.enum([
        "light",
        "dark"
    ]),
    accentColor: _zod.z.string(),
    dangerColor: _zod.z.string(),
    grayColor: _zod.z.string(),
    infoColor: _zod.z.string(),
    successColor: _zod.z.string(),
    warningColor: _zod.z.string()
}).partial();
